### PR TITLE
Add v4 dynamic validation checks and tests

### LINKED ISSUE #69: Adapt dynamic config validation to conform to v4 schema
- must validate `derived_task_ids` are valid task IDs
- Any dynamic checks must take into account that `optional` properties are no longer supplied in `output_type_id`s

### PR DESCRIPTION
**Pull request by annakrystalli**
This PR resolves #69 by adding dynamic validation checks to ensure derived task ID match valid task ID names at the config and/or round level.

All other v4 schema changes are captured by `jsonvalidate` validation.

The PR also adds test v4 task.json config files and tests.

I've also minorly refactored and added documentation to the `get_error_path()`. Also, now when multiple matching paths to the required property are found (which has now been introduced with the config and round level `derived_task_ids` properties), the function returns the first match at the highest depth.

### CODE REVIEW

File: R/get_error_path.R [hunk 1]
```diff
@@ -70,23 +103,35 @@ get_error_path <- function(schema, element = "target_metadata",
...
     unlist()
-
   is_item <- split_path == "items"
-  split_path[is_item] <- c(
-    "{round_i - 1}",
-    "{model_task_i - 1}",
-    "{target_key_i - 1}"
-  )[1:sum(is_item)]
+  n_item_props <- sum(is_item)
+  split_path[is_item] <- utils::head(item_props, n_item_props)
```
Comment (zkamvar) [inline R/get_error_path.R:126]:
```suggestion
  split_path[is_item] <- item_props[seq_len(n_item_props)]
```

subsetting feels more natural and is faster by an order of magnitude.

``` r
x <- 1:10
bench::mark(utils::head(x, 3), x[seq_len(3)])
#> # A tibble: 2 × 6
#>   expression             min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>        <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 utils::head(x, 3)   1.52µs   1.68µs   511741.        0B     51.2
#> 2 x[seq_len(3)]         41ns     82ns  8051712.        0B      0
```

<sup>Created on 2024-11-06 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

File: R/get_error_path.R [hunk 2]
```diff
@@ -70,23 +103,35 @@ get_error_path <- function(schema, element = "target_metadata",
...
   paste(split_path, collapse = "/")
 }
+
+# Find the first path in a vector of paths by depth (ordered by order of depth -
+# deepest last)
+detect_first <- function(paths) {
+  min_idx <- stringr::str_split(paths, "/") |>
+    lengths() |>
+    which.min()
+  paths[min_idx]
```
Comment (zkamvar) [inline R/get_error_path.R:136]: Clever solution---I like it!

File: R/get_error_path.R [hunk 3]
```diff
@@ -59,6 +87,11 @@ get_error_path <- function(schema, element = "target_metadata",
       unique()
   }
 
+  # If the path we are targeting is an an array (i.e. property name is followed
+  # by "items" and `append_item_n` is `TRUE`, (i.e. we want to include the
+  # index to a specific element in that object), then add an "\items" string to the
+  # end of the path where a location index will be added during instance
+  # path interpolation.
```
Comment (zkamvar) [inline R/get_error_path.R:94]: 🙏🏽 thank you!

Comment (zkamvar) [reply R/get_error_path.R:94]: One suggestion I have would be to move out the subsequent validation checks to sentinel variables above the if statement. For example:

```r
  path_exists <- length(path) != 0
  is_array <- path_exists && any(grepl(paste(path, "items", sep = "/"), schema_paths))
  if (append_item_n && is_array) {
```

**Review by zkamvar — APPROVED** [review]
Thank you! Again, I appreciate the context you provide in the comments. It's good to merge with or without the suggestions.

File: R/get_error_path.R (hunk 1 above)
Comment (annakrystalli) [reply R/get_error_path.R:126]: These really are tiny performance improvements when we're talking about a length 3 vector.
More importantly, indexing returns an error when `n_item_props` is 0L while `head()` does not
``` r
1:3[seq_len(0L)]
#> Error in 1:3[seq_len(0L)]: argument of length 0

head(1:3, 0L)
#> integer(0)
```
I also find `head()` much clearer

Comment (annakrystalli) [reply R/get_error_path.R:126]: Actually the above is a problem with subsetting a `:` expression. Subsetting with a zero length works for already defined objects, so I've gone ahead and changed it.

``` r
paths <- vector(length = 3L)
n <- sum(paths)
paths[seq_len(n)]
#> logical(0)
```

<sup>Created on 2024-11-07 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

File: R/get_error_path.R (hunk 3 above)
Comment (annakrystalli) [reply R/get_error_path.R:94]: Done

**Pull request merged**